autoware_lanelet2_extension/include/autoware_lanelet2_extension/io/autoware_multi_osm_parser.hpp

```diff
@@ -0,0 +1,188 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef AUTOWARE_LANELET2_EXTENSION__IO__AUTOWARE_MULTI_OSM_PARSER_HPP_
+#define AUTOWARE_LANELET2_EXTENSION__IO__AUTOWARE_MULTI_OSM_PARSER_HPP_
+
+#include "lanelet2_io/Exceptions.h"
+#include "lanelet2_io/io_handlers/Factory.h"
+#include "lanelet2_io/io_handlers/OsmFile.h"
+#include "lanelet2_io/io_handlers/OsmHandler.h"
+
+#include <boost/geometry/algorithms/is_valid.hpp>
+
+#include <lanelet2_core/geometry/LineString.h>
+#include <lanelet2_core/geometry/Polygon.h>
+
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+using std::string_literals::operator""s;
+
+namespace lanelet
+{
+namespace io_handlers
+{
+
+class MultiOsmParser : public Parser
+{
+public:
+  using Parser::Parser;
+
+  std::unique_ptr<LaneletMap> parse(
+    const std::string & lanelet2_filename, ErrorMessages & errors) const override;
+
+  std::unique_ptr<LaneletMap> parse(
+    const std::vector<std::string> & lanelet2_filenames, ErrorMessages & errors) const;
+
+  std::unique_ptr<LaneletMap> fromOsmFile(const osm::File & file, ErrorMessages & errors) const;
+
+  std::unique_ptr<LaneletMap> fromOsmFile(
+    const std::map<std::string, osm::File> & files_map, ErrorMessages & errors) const;
+
+  static void parseVersions(
+    const std::string & filename, std::string * format_version, std::string * map_version);
+
+  static constexpr const char * extension() { return ".osm"; }
+
+  static constexpr const char * name() { return "autoware_multi_osm_handler"; }
+};
+
+namespace
+{
+RegisterParser<MultiOsmParser> regParser;
+
+class MultiFileLoader
+{
+public:
+  static std::unique_ptr<LaneletMap> loadMap(
+    const osm::File & file, const Projector & projector, ErrorMessages & errors)
+  {
+    MultiFileLoader loader;
+    loader.loadNodes(file.nodes, projector);
+    loader.loadWays(file.ways);
+    auto laneletsWithRelation = loader.loadLanelets(file.relations);
+    auto areasWithRelation = loader.loadAreas(file.relations);
+    loader.loadRegulatoryElements(file.relations);
+    loader.addRegulatoryElements(laneletsWithRelation);
+    loader.addRegulatoryElements(areasWithRelation);
+    errors = std::move(loader.errors_);
+
+    return std::make_unique<LaneletMap>(
+      loader.lanelets_, loader.areas_, loader.regulatoryElements_, loader.polygons_,
+      loader.lineStrings_, loader.points_);
+  }
+
+  static std::unique_ptr<LaneletMap> loadMap(
```

Comment:
Is this future related to the differential loading mechanism ? Then can we dynamically add next map while dropping visited map by using the `add` and `removed` function to each layers, so we do not need to recreate the entire lanelet map whenever ego passes the boundary ?

https://github.com/fzi-forschungszentrum-informatik/Lanelet2/blob/507033b82d9915f086f2539d56c3b62e71802438/lanelet2_core/include/lanelet2_core/LaneletMap.h#L245-L263